File: frontend/js/dividendos.js
async function carregarDividendos() {
  const tbody = document.getElementById('tabela-dividendos');
  const resumo = document.getElementById('resumo-dividendos');

  try {
    const [divs, ativos] = await Promise.all([
      fetch('/dividendos').then(r => r.json()),
      fetch('/ativos').then(r => r.json())
    ]);

    // Resumo total
    const totalRecebido = divs.reduce((sum, d) => sum + d.valor_total, 0);
    const mesAtual = new Date().toISOString().slice(0, 7);
    const totalMes = divs.filter(d => d.data_pagamento.startsWith(mesAtual))
                        .reduce((sum, d) => sum + d.valor_total, 0);

    resumo.innerHTML = `
      <div class="grid-4" style="margin-bottom:20px;">
        <div class="card">
          <div class="card-label">Total Recebido</div>
          <div class="card-value green">${fmtMoeda(totalRecebido)}</div>
          <div class="card-sub">${divs.length} proventos registrados</div>
        </div>
        <div class="card">
          <div class="card-label">Recebido este mês</div>
          <div class="card-value">${fmtMoeda(totalMes)}</div>
          <div class="card-sub pos">mês atual</div>
        </div>
        <div class="card">
          <div class="card-label">Média Mensal</div>
          <div class="card-value">${fmtMoeda(divs.length > 0 ? totalRecebido / 12 : 0)}</div>
          <div class="card-sub">estimativa anualizada</div>
        </div>
        <div class="card">
          <div class="card-label">Ativos com Proventos</div>
          <div class="card-value">${new Set(divs.map(d => d.ticker)).size}</div>
          <div class="card-sub">tickers distintos</div>
        </div>
      </div>
    `;

    if (!divs.length) {
      tbody.innerHTML = '<tr><td colspan="6" class="loading">Nenhum provento registrado.</td></tr>';
      return;
    }

    tbody.innerHTML = divs.map(d => `<tr>
      <td><span style="font-weight:700;">${d.ticker}</span></td>
      <td><span class="pill p-blue">${d.tipo}</span></td>
      <td>${fmtMoeda(d.valor_por_cota)}/cota</td>
      <td>${fmt(d.quantidade_cotas, 0)} cotas</td>
      <td style="font-weight:700;color:#1E6E3A;">${fmtMoeda(d.valor_total)}</td>
      <td>${new Date(d.data_pagamento).toLocaleDateString('pt-BR')}</td>
    </tr>`).join('');

  } catch(e) {
    tbody.innerHTML = '<tr><td colspan="6" class="loading">Erro ao carregar.</td></tr>';
  }
}

async function carregarRetornoTotal() {
  const el = document.getElementById('retorno-total-lista');
  el.innerHTML = '<div class="loading">Calculando retorno total...</div>';

  try {
    const ativos = await fetch('/ativos').then(r => r.json());
    const fiis = ativos.filter(a => a.classe === 'FII');

    if (!fiis.length) {
      el.innerHTML = '<div class="loading">Nenhum FII na carteira.</div>';
      return;
    }

    const resultados = await Promise.all(
      fiis.map(a => fetch(`/retorno-total/${a.ticker}`).then(r => r.json()))
    );

    let html = `
      <div class="table-wrap">
        <table>
          <thead><tr>
            <th>Ativo</th>
            <th>Valoriz. Cota</th>
            <th>Dividendos Recebidos</th>
            <th>DY s/ PM</th>
            <th>Retorno Total</th>
            <th>Proventos</th>
          </tr></thead>
          <tbody>
    `;

    resultados.forEach(r => {
      const totCls = r.retorno_total_pct >= 0 ? 'p-green' : 'p-red';
      const cotaCls = r.retorno_cota_pct >= 0 ? 'pos' : 'neg';
      html += `<tr>
        <td><span style="font-weight:700;">${r.ticker}</span></td>
        <td>
          <div class="${cotaCls}" style="font-weight:600;">${r.retorno_cota_pct >= 0 ? '+' : ''}${fmt(r.retorno_cota_pct)}%</div>
          <div style="font-size:10px;color:#888;">${fmtMoeda(r.retorno_cota_rs)}</div>
        </td>
        <td style="color:#1E6E3A;font-weight:600;">${fmtMoeda(r.total_dividendos)}</td>
        <td style="color:#1E6E3A;font-weight:600;">+${fmt(r.dy_recebido_pct)}%</td>
        <td><span class="pill ${totCls}" style="font-size:13px;padding:4px 10px;">+${fmt(r.retorno_total_pct)}%</span>
          <div style="font-size:10px;color:#888;margin-top:3px;">${fmtMoeda(r.retorno_total_rs)}</div>
        </td>
        <td style="color:#888;">${r.qtd_proventos} registros</td>
      </tr>`;
    });

    html += '</tbody></table></div>';
    el.innerHTML = html;

  } catch(e) {
    el.innerHTML = '<div class="loading">Erro ao calcular retorno total.</div>';
  }
}

async function registrarDividendo() {
  const msg = document.getElementById('msg-dividendo');
  const ticker = document.getElementById('div-ticker').value.trim().toUpperCase();
  const valor_por_cota = parseFloat(document.getElementById('div-valor-cota').value);
  const quantidade_cotas = parseFloat(document.getElementById('div-quantidade').value);
  const data_pagamento = document.getElementById('div-data').value;
  const tipo = document.getElementById('div-tipo').value;

  if (!ticker || !valor_por_cota || !quantidade_cotas || !data_pagamento) {
    msg.innerHTML = '<div class="alert alert-red">Preencha todos os campos.</div>';
    return;
  }

  try {
    const res = await fetch('/dividendos', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ ticker, valor_por_cota, quantidade_cotas, data_pagamento, tipo })
    });
    const data = await res.json();
    if (res.ok) {
      msg.innerHTML = `<div class="alert alert-green">✓ ${data.mensagem} · Total: ${fmtMoeda(data.valor_total)}</div>`;
      document.getElementById('div-ticker').value = '';
      document.getElementById('div-valor-cota').value = '';
      document.getElementById('div-quantidade').value = '';
      document.getElementById('div-data').value = '';
      carregarDividendos();
      carregarRetornoTotal();
    } else {
      msg.innerHTML = `<div class="alert alert-red">✗ ${data.detail}</div>`;
    }
  } catch(e) {
    msg.innerHTML = '<div class="alert alert-red">Erro ao registrar.</div>';
  }
}

async function importarTodos() {
  const msg = document.getElementById('msg-dividendo');
  msg.innerHTML = '<div class="alert alert-green">⏳ Importando proventos de todos os ativos...</div>';

  try {
    const res = await fetch('/dividendos/importar-todos', { method: 'POST' });
    const data = await res.json();

    if (data.resultados && data.resultados.length > 0) {
      const detalhes = data.resultados.map(r => `${r.ticker}: ${r.importados} proventos`).join(' · ');
      msg.innerHTML = `<div class="alert alert-green">✓ Importação concluída — ${detalhes}</div>`;
    } else {
      msg.innerHTML = '<div class="alert alert-green">✓ Nenhum provento novo encontrado. Tudo já estava atualizado.</div>';
    }

    carregarDividendos();
    carregarRetornoTotal();
  } catch(e) {
    msg.innerHTML = '<div class="alert alert-red">Erro na importação.</div>';
  }
}

async function importarTicker() {
  const ticker = document.getElementById('import-ticker').value.trim();
  const msg = document.getElementById('msg-dividendo');

  if (!ticker) {
    msg.innerHTML = '<div class="alert alert-red">Digite um ticker.</div>';
    return;
  }

  msg.innerHTML = `<div class="alert alert-green">⏳ Importando proventos de ${ticker}...</div>`;

  try {
    const res = await fetch(`/dividendos/importar/${ticker}`, { method: 'POST' });
    const data = await res.json();

    if (res.ok) {
      msg.innerHTML = `<div class="alert alert-green">✓ ${data.mensagem}</div>`;
      document.getElementById('import-ticker').value = '';
      carregarDividendos();
      carregarRetornoTotal();
    } else {
      msg.innerHTML = `<div class="alert alert-red">✗ ${data.detail}</div>`;
    }
  } catch(e) {
    msg.innerHTML = '<div class="alert alert-red">Erro na importação.</div>';
  }
}

File: backend/routers/dividendos.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import desc, func
from pydantic import BaseModel
from datetime import date
import yfinance as yf

from backend.database import get_db, Ativo, Dividendo
from backend.data.cache import buscar_preco_com_cache as buscar_preco

router = APIRouter()


class DividendoCreate(BaseModel):
    ticker: str
    valor_por_cota: float
    quantidade_cotas: float
    data_pagamento: date
    tipo: str = "PROVENTO"


@router.post("/dividendos")
def registrar_dividendo(div: DividendoCreate, db: Session = Depends(get_db)):
    ticker = div.ticker.upper()
    valor_total = div.valor_por_cota * div.quantidade_cotas
    db.add(Dividendo(
        ticker=ticker,
        valor_por_cota=div.valor_por_cota,
        quantidade_cotas=div.quantidade_cotas,
        valor_total=valor_total,
        data_pagamento=div.data_pagamento,
        tipo=div.tipo,
    ))
    db.commit()
    return {"mensagem": f"Dividendo de {ticker} registrado", "valor_total": round(valor_total, 2)}


@router.get("/dividendos/{ticker}")
def listar_dividendos(ticker: str, db: Session = Depends(get_db)):
    items = db.query(Dividendo).filter(Dividendo.ticker == ticker.upper()).order_by(desc(Dividendo.data_pagamento)).all()
    return [{"id": i.id, "ticker": i.ticker, "valor_por_cota": i.valor_por_cota,
             "quantidade_cotas": i.quantidade_cotas, "valor_total": i.valor_total,
             "data_pagamento": str(i.data_pagamento), "tipo": i.tipo} for i in items]


@router.get("/dividendos")
def listar_todos_dividendos(db: Session = Depends(get_db)):
    items = db.query(Dividendo).order_by(desc(Dividendo.data_pagamento)).all()
    return [{"id": i.id, "ticker": i.ticker, "valor_por_cota": i.valor_por_cota,
             "quantidade_cotas": i.quantidade_cotas, "valor_total": i.valor_total,
             "data_pagamento": str(i.data_pagamento), "tipo": i.tipo} for i in items]


@router.get("/retorno-total/{ticker}")
async def retorno_total(ticker: str, db: Session = Depends(get_db)):
    ticker = ticker.upper()
    ativo = db.query(Ativo).filter(Ativo.ticker == ticker, Ativo.ativo == True).first()
    if not ativo:
        raise HTTPException(status_code=404, detail="Ativo não encontrado")

    result = db.query(func.sum(Dividendo.valor_total), func.count(Dividendo.id)).filter(Dividendo.ticker == ticker).one()
    total_dividendos = result[0] or 0
    qtd_proventos = result[1] or 0

    preco_atual = await buscar_preco(ativo.ticker, ativo.mercado)
    preco = preco_atual.get("preco") or ativo.preco_medio

    valor_atual = preco * ativo.quantidade
    valor_investido = ativo.preco_medio * ativo.quantidade
    retorno_cota = valor_atual - valor_investido
    retorno_cota_pct = (retorno_cota / valor_investido * 100) if valor_investido > 0 else 0
    dy_pct = (total_dividendos / valor_investido * 100) if valor_investido > 0 else 0

    return {
        "ticker": ticker,
        "quantidade": ativo.quantidade,
        "preco_medio": ativo.preco_medio,
        "preco_atual": preco,
        "valor_investido": round(valor_investido, 2),
        "valor_atual": round(valor_atual, 2),
        "retorno_cota_rs": round(retorno_cota, 2),
        "retorno_cota_pct": round(retorno_cota_pct, 2),
        "total_dividendos": round(total_dividendos, 2),
        "dy_recebido_pct": round(dy_pct, 2),
        "retorno_total_rs": round(retorno_cota + total_dividendos, 2),
        "retorno_total_pct": round(retorno_cota_pct + dy_pct, 2),
        "qtd_proventos": qtd_proventos,
    }


@router.post("/dividendos/importar/{ticker}")
def importar_dividendos(ticker: str, db: Session = Depends(get_db)):
    ticker = ticker.upper()
    ativo = db.query(Ativo).filter(Ativo.ticker == ticker, Ativo.ativo == True).first()
    if not ativo:
        raise HTTPException(status_code=404, detail="Ativo não encontrado na carteira")

    try:
        ticker_yf = f"{ticker}.SA" if ativo.mercado == "BR" else ticker
        divs = yf.Ticker(ticker_yf).dividends
        if divs.empty:
            return {"mensagem": "Nenhum dividendo encontrado", "importados": 0}

        if ativo.data_compra:
            divs = divs[divs.index >= str(ativo.data_compra)]

        importados, ignorados = 0, 0
        for data, valor in divs.items():
            data_obj = date.fromisoformat(str(data)[:10])
            existente = db.query(Dividendo).filter(
                Dividendo.ticker == ticker,
                Dividendo.data_pagamento == data_obj,
            ).first()
            if existente:
                ignorados += 1
                continue
            db.add(Dividendo(
                ticker=ticker,
                valor_por_cota=float(valor),
                quantidade_cotas=ativo.quantidade,
                valor_total=float(valor) * ativo.quantidade,
                data_pagamento=data_obj,
                tipo="AUTO",
            ))
            importados += 1

        db.commit()
        return {
            "mensagem": f"{ticker} — {importados} proventos importados",
            "importados": importados,
            "ignorados": ignorados,
            "quantidade_cotas": ativo.quantidade,
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/dividendos/importar-todos")
def importar_todos_dividendos(db: Session = Depends(get_db)):
    ativos = db.query(Ativo).filter(Ativo.ativo == True).all()
    resultados = []
    for ativo in ativos:
        try:
            ticker_yf = f"{ativo.ticker}.SA" if ativo.mercado == "BR" else ativo.ticker
            divs = yf.Ticker(ticker_yf).dividends
            if divs.empty:
                continue
            if ativo.data_compra:
                divs = divs[divs.index >= str(ativo.data_compra)]
            importados = 0
            for data, valor in divs.items():
                data_obj = date.fromisoformat(str(data)[:10])
                existente = db.query(Dividendo).filter(
                    Dividendo.ticker == ativo.ticker,
                    Dividendo.data_pagamento == data_obj,
                ).first()
                if existente:
                    continue
                db.add(Dividendo(
                    ticker=ativo.ticker,
                    valor_por_cota=float(valor),
                    quantidade_cotas=ativo.quantidade,
                    valor_total=float(valor) * ativo.quantidade,
                    data_pagamento=data_obj,
                    tipo="AUTO",
                ))
                importados += 1
            db.commit()
            if importados > 0:
                resultados.append({"ticker": ativo.ticker, "importados": importados})
        except Exception:
            continue
    return {"mensagem": "Importação concluída", "resultados": resultados}
